### ISSUE TITLE
Clean up gopt_f77_methods.f90

### ISSUE THREAD

**Issue by oschuett** [cp2k/cp2k#1233]
The [gopt_f77_methods.fypp](https://github.com/cp2k/cp2k/blob/master/src/motion/gopt_f77_methods.fypp) is a hack to implement a circular dependency between modules. It's needed for the `GEO_OPT` and `MD` type of [cell optimization](https://manual.cp2k.org/trunk/CP2K_INPUT/MOTION/CELL_OPT.html#TYPE). They are both obsolete and we should simply remove them to eliminate that circular dependency.

**Comment by dev-zero** [comment]
see also #1231

**Comment by oschuett** [comment]
Upps, I overlook that. Should I close this issue or shall we specialize them and make  #1231 about `f77_interface.F` only?

**Comment by dev-zero** [comment]
Probably specializing them makes sense, this way we can (hopefully) close issues directly with PRs and we get a better overview rather than having subtasks in issues. I wasn't aware that the f77 interface in `motion/` is actually more a workaround rather than an API. Thanks!

**Comment by he-zilong** [comment]
A major distinction between options of `&MOTION/&CELL_OPT/TYPE` is how the pressure tensor is calculated which is handled by several `BLOCK`s in `cp_eval_at()` of `gopt_f77_methods.F`. Why this design? Would a refactor where these `BLOCK`s are converted to proper `SUBROUTINE`s in a separate `MODULE` help with this issue?

**Comment by oschuett** [comment]
I think, we should just remove the [&MOTION/&CELL_OPT/TYPE](https://manual.cp2k.org/trunk/CP2K_INPUT/MOTION/CELL_OPT.html#TYPE) keyword. AFAIK, it's not used and without it we don't have the recursive call path that leads to the circular dependency between modules.

**Comment by he-zilong** [comment]
> I think, we should just remove the [&MOTION/&CELL_OPT/TYPE](https://manual.cp2k.org/trunk/CP2K_INPUT/MOTION/CELL_OPT.html#TYPE) keyword. AFAIK, it's not used and without it we don't have the recursive call path that leads to the circular dependency between modules.

The [manual description](https://manual.cp2k.org/cp2k-2026_1-branch/CP2K_INPUT/MOTION/CELL_OPT.html#CP2K_INPUT.MOTION.CELL_OPT.TYPE) (link is to the last stable release; although I made some revisions in #5086 there was no substantial change in the meaning and implementation) seemed to suggest that the options other than `DIRECT_CELL_OPT` might be useful for reducing computation cost of stress tensor evaluation, as pointed out in #4426 where "analytical stress tensor is unavailable" for r2SCAN-D4.

**Comment by oschuett** [comment]
> ... as pointed out in https://github.com/cp2k/cp2k/issues/4426 where "analytical stress tensor is unavailable" for r2SCAN-D4.

Thanks for the pointer! Still, I'm not convinced this proofs significant usage of other cell opt types besides `DIRECT_CELL_OPT`.

AFAIK, we do have analytic stress tensors for almost all methods. @mkrack, what do you think?

**Comment by mkrack** [comment]
The CELL_OPT type MD is computationally very expensive and converging very badly due to noise. I don't know a use case for this CELL_OPT type which could not be achieved also with the RUN_TYPE MD and NpT in a better way.
I also never found a case for which the CELL OPT type GEO_OPT was more efficient than the DIRECT_CELL_OPT type. The type GEO_OPT became quasi obsolete after the DIRECT_CELL_OPT type was implemented, but I never compared the performance of these types using a numerical stress tensor.

**Comment by oschuett** [comment]
Unfortunately, even with [CELL_OPT/TYPE removed](https://github.com/oschuett/cp2k/tree/gopt_f77_methods) there remains a circular dependency:

The Dimer method runs a nested [ROT_OPT](https://manual.cp2k.org/trunk/CP2K_INPUT/MOTION/GEO_OPT/TRANSITION_STATE/DIMER/ROT_OPT.html) inside the outer translation optimization.

I'm not sure what's the best way to disentangle this.

**Comment by he-zilong** [comment]
Yeah, extending support to analytical stress tensors (especially popular DFT methods, perhaps even including TDDFT excited states) is the way to go.

I heard that the dimer method supports only `CG`/`2PNT` for both translation and rotation optimizations. Is this true? Is it possible to replace the nested optimization procedures with a single one, where a flag marks whether each step is optimizing translation or rotation and toggles according to convergence criteria?

**Comment by oschuett** [comment]
> I heard that the dimer method supports only CG/2PNT for both translation and rotation optimizations. Is this true?

Yes, the outer optimization works only with CG.

>  Is it possible to replace the nested optimization procedures with a single one,...

Maybe. It would probably mean loosing the state of the outer optimization, which would turn it into steepest descent.

**Comment by he-zilong** [comment]
> > Is it possible to replace the nested optimization procedures with a single one,...
> 
> Maybe. It would probably mean loosing the state of the outer optimization, which would turn it into steepest descent.

What about preserving the algorithm choice for translation (outer) optimization but reducing the algorithm choice for rotation (inner) optimization to steepest descent only? Given a not-too-large step size, the dimer does not move very far on the potential surface in a translation (outer) optimization step and its orientation does not change much, so the distinction between algorithms for rotation (inner) optimization may not be that significant.

IMHO, the one crucial ingredient for a transition state search is really the initial dimer vector provided alongside the initial structure. While both (1) a normal mode from a vibrational analysis job and (2) the difference between two second-highest points from a band (CI-NEB) job are a good ansatz for [`&MOTION/&GEO_OPT/&TRANSITION_STATE/&DIMER/&DIMER_VECTOR`](https://manual.cp2k.org/trunk/CP2K_INPUT/MOTION/GEO_OPT/TRANSITION_STATE/DIMER/DIMER_VECTOR.html), they may not be favorable in terms of computational cost. Thus we need a cheaper generator for an initial guess of dimer vector that is also better than the current random one.

**Comment by he-zilong** [comment]
> Unfortunately, even with [CELL_OPT/TYPE removed](https://github.com/oschuett/cp2k/tree/gopt_f77_methods) there remains a circular dependency...

Maybe the removal of `&CELL_OPT/TYPE` keyword itself (without resolving the circular dependency, so still keeping the `gopt_f77_methods.F` and `gopt_f77_methods.fypp` workaround) should be submitted as a standalone PR first as it is already a "breaking change" to be mentioned in the next release. This way #4426 could be closed for the moment, and the implementation of analytical stress tensor for MGGA functionals or DFT-D4 corrections would be relegated to separate feature request(s) (if not already available nowadays, that is; still need verification). And then move on to the circular dependency caused by `ROT_OPT`...

Edit: I would like to work on the initial dimer vector in a few days after the storms pass and regtests are back in working order.

**Comment by oschuett** [comment]
>  I would like to work on the initial dimer vector...

I'm not sure if this is helpful, but we do have a [model hessian](https://manual.cp2k.org/trunk/CP2K_INPUT/MOTION/CELL_OPT/BFGS.html#CP2K_INPUT.MOTION.CELL_OPT.BFGS.USE_MODEL_HESSIAN) for initializing BFGS.

https://github.com/cp2k/cp2k/blob/15814972975c5bb13e63340cddf430d6296ccfd1/src/motion/bfgs_optimizer.F#L1027

**Comment by he-zilong** [comment]
> I'm not sure if this is helpful, but we do have a [model hessian](https://manual.cp2k.org/trunk/CP2K_INPUT/MOTION/CELL_OPT/BFGS.html#CP2K_INPUT.MOTION.CELL_OPT.BFGS.USE_MODEL_HESSIAN) for initializing BFGS.

Thank you for the reminder. Incorporating this subroutine is theoretically feasible but perhaps a little overkill for the purpose; after all, a Hessian matrix can be converted to a force constant matrix and then diagonalized to give 3N vibrational normal modes, whereas the initial dimer vector would need only 1 normal mode. Moreover, the automatic construction of the model Hessian probably means that the user does not have a chance to visualize the resulting normal modes and pick one with desirable atomic motion as close to the target transition state as possible. What I have in mind is different...

Where do these magic constants come from, anyway?
https://github.com/cp2k/cp2k/blob/15814972975c5bb13e63340cddf430d6296ccfd1/src/motion/bfgs_optimizer.F#L1049-L1055

**Comment by oschuett** [comment]
> Where do these magic constants come from, anyway?

While it's not cited anywhere in the code, I believe it's [this method](https://doi.org/10.1016/0009-2614(95)00646-L).

**Comment by RTChou67** [comment]
How about moving `cp_eval_at` to the `gopt_f_methods` module, and remove `gopt_f77_methods.F` and `gopt_f77_methods.fypp`? For the circular dependencies, how about adding an ABSTRACT INTERFACE to `gopt_f_type` and a corresponding procedure-pointer component to `gopt_f_type`. Then `geo_opt` could assign `cp_rot_opt` to this procedure pointer, while dimer methods could call the rotation optimizer through the pointer without directly depending on `geo_opt`.

**Comment by oschuett** [comment]
>  For the circular dependencies, how about adding an ABSTRACT INTERFACE to gopt_f_type and a corresponding procedure-pointer component to gopt_f_type.

Yes, that's actually a great idea!
